### PR TITLE
Fixes #26772: FileField should re-open closed files

### PR DESCRIPTION
**Pull request by pjstevns**
fix a regression from refactoring in a52a531a8b34f049fba11c3ee7b010af7534bf90

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/file_storage/tests.py [hunk 1]
```diff
@@ -746,6 +746,15 @@ def test_filefield_write(self):
         self.assertEqual(obj.normal.read(), b'updated')
         obj.normal.close()
 
+    def test_filefield_reopen(self):
+        # Files can be written to.
```
Comment (bmispelon) [inline tests/file_storage/tests.py]: Could you update the comment here to reflect what the test is doing?

**Comment by charettes** [comment]
Are you sure this is a regression in Django 1.10?

While I think this is a desirable behavior, from my personal experience `FieldFile` were never re-opened in the past. As a matter of fact I ran your test against Django 1.9 without your proposed fix and it also failed.

File: tests/file_storage/tests.py [hunk 2]
```diff
@@ -746,6 +746,15 @@ def test_filefield_write(self):
...
+        obj = Storage.objects.create(normal=SimpleUploadedFile('rewritten.txt', b'content'))
+        with obj.normal as normal:
+            normal.open('wb')
```
Comment (charettes) [inline tests/file_storage/tests.py]: I would reduce the test to a simpler expression:

```python
with obj.normal as normal:
    obj.normal.open()
obj.normal.open()
obj.normal.file.seek(0)
```

**Comment by charettes** [comment]
This change makes me wonder what should be done  if `_file` is already assigned and not closed but opened in a different mode. Before a52a531a8b34f049fba11c3ee7b010af7534bf90 this was not an issue as new file was returned but now the following fails:

```python
# Somewhere the `FileField` descriptor is accessed and `_file` is assigned
obj.file_field.read()  # This opens `_file` in 'rb' mode without closing it.
...
# Attempt to open in `wb` mode to alter the file.
obj.file_field.open('wb')
obj.file_field.wirte(b'foo')  # This won't work as `_file` is still opened in `rb` mode.
```

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
I was thinking of adding something like `or (hasattr(self._file,'mode') and self._file.mode != mode)` to the condition in `_get_file()` that reopens the file, but that causes accessing the file property `file = property(_get_file, ...)` to always change things to `rb` mode.

Here's the test I was using based on Simon's comment:
```python
def test_filefield_reopen_no_close(self):
    obj = Storage.objects.create(normal=SimpleUploadedFile('rewritten.txt', b'content'))
    obj.normal.read()
    obj.normal.open('wb')
    obj.normal.file.write(b'foo')
```
Maybe it's best to revert the patch that caused the regression for now.

File: tests/file_storage/tests.py (hunk 2 above)
Comment (timgraham) [inline tests/file_storage/tests.py]: And add `obj.normal.close()` at the end to fix the failure on Windows.

**Comment by charettes** [comment]
> Maybe it's best to revert the patch that caused the regression for now.

Agreed, this change needs some more thought.

**Comment by timgraham** [comment]
Original fix is reverted and here's a new PR to add the tests suggested by this ticket: https://github.com/django/django/pull/6797.

**Pull request closed without merging**